plugins/osdn/vnids.go

```diff
@@ -9,22 +9,57 @@ import (
 
 	kapi "k8s.io/kubernetes/pkg/api"
 	kubetypes "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/registry/service/allocator"
+	etcdallocator "k8s.io/kubernetes/pkg/registry/service/allocator/etcd"
+	"k8s.io/kubernetes/pkg/util"
 	kerrors "k8s.io/kubernetes/pkg/util/errors"
 	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
-	"k8s.io/kubernetes/pkg/util/sets"
 	utilwait "k8s.io/kubernetes/pkg/util/wait"
 	"k8s.io/kubernetes/pkg/watch"
 
-	"github.com/openshift/openshift-sdn/pkg/netutils"
-	osapi "github.com/openshift/origin/pkg/sdn/api"
+	"github.com/openshift/openshift-sdn/pkg/netid"
+	vnidcontroller "github.com/openshift/openshift-sdn/plugins/osdn/netid/controller"
+	"github.com/openshift/openshift-sdn/plugins/osdn/netid/vnid"
+	"github.com/openshift/openshift-sdn/plugins/osdn/netid/vnidallocator"
+	"github.com/openshift/origin/pkg/sdn/registry/netnamespace"
 )
 
-const (
-	// Maximum VXLAN Network Identifier as per RFC#7348
-	MaxVNID = ((1 << 24) - 1)
-	// VNID for the admin namespaces
-	AdminVNID = uint(0)
-)
+func (oc *OsdnController) VnidStartMaster() error {
+	netIDRange, err := vnid.NewVNIDRange(netid.MinVNID, netid.MaxVNID-netid.MinVNID+1)
+	if err != nil {
+		return fmt.Errorf("Unable to create NetID range: %v", err)
+	}
+
+	var etcdAlloc *etcdallocator.Etcd
+	netIDAllocator := vnidallocator.New(netIDRange, func(max int, rangeSpec string) allocator.Interface {
+		mem := allocator.NewContiguousAllocationMap(max, rangeSpec)
+		etcdAlloc = etcdallocator.NewEtcd(mem, "/ranges/namespacevnids", kapi.Resource("namespacevnidallocation"), oc.EtcdHelper)
+		return etcdAlloc
+	})
+
+	_, kclient := oc.Registry.GetSDNClients()
+
+	// Run vnid migration
+	migrate := netnamespace.NewMigrate(oc.EtcdHelper, kclient.Namespaces())
+	if err := migrate.Run(); err != nil {
+		return fmt.Errorf("Unable to migrate network namespaces: %v", err)
```

Comment:
> If the problem persists (very rare) they could delete NetNamespace objects manually using oc delete or curl on etcd

they can't use `oc delete` if the master won't start, right? or does this kill controllers but not the master API?

> Migration order for master or node doesn't matter but all the nodes need to be migrated as well. Old nodes will watch for NetNamespace and this resource no longer exists. 

If I understand correctly, new nodes will only pay attention to namespace annotations, and old nodes will only pay attention to NetNamespace objects. As written, migration removes old objects as it creates the annotations, which means there is no upgrade order that ensures all actors have the information they need:
* upgrading nodes first means new nodes will be missing annotations on namespace until migration runs
* upgrading masters first means old nodes will miss NetNamespace objects until upgraded

I don't know what the nodes do with network setup when they are missing configured vnids, but it seems like it would be disruptive...